pkg/apply/destroyer.go

```diff
@@ -68,107 +80,62 @@ func (d *Destroyer) Initialize() error {
-		// it should terminate.
-		close(tempChannel)
-		// Wait for the event transformer to complete processing all
-		// events and shut down before we continue.
-		<-completedChannel
-		ch <- event.Event{
-			Type: event.ActionGroupType,
-			ActionGroupEvent: event.ActionGroupEvent{
-				GroupName: "destroyer",
-				Type:      event.Finished,
-				Action:    event.DeleteAction,
+		resourceObjs := &ResourceObjects{LocalInv: inv, PruneIds: pruneIds}
+		klog.V(4).Infoln("destroyer building task queue...")
+		taskBuilder := &solver.TaskQueueBuilder{
+			PruneOptions: d.PruneOptions,
+			Factory:      d.provider.Factory(),
+			Mapper:       mapper,
+			InvClient:    d.invClient,
+		}
+		opts := solver.Options{
+			Prune:                  true,
+			PruneTimeout:           time.Duration(0),
+			DryRunStrategy:         option.DryRunStrategy,
+			PrunePropagationPolicy: metav1.DeletePropagationBackground,
+			InventoryPolicy:        option.InventoryPolicy,
+			ReconcileTimeout:       time.Duration(0),
+		}
+		// Build the ordered set of tasks to execute.
+		taskQueue := taskBuilder.
+			BuildPruneTaskQueue(resourceObjs, opts).
+			AppendDeleteInvTask(resourceObjs).
+			Build()
+		// Send event to inform the caller about the resources that
+		// will be pruned.
+		eventChannel <- event.Event{
+			Type: event.InitType,
+			InitEvent: event.InitEvent{
+				ActionGroups: taskQueue.ToActionGroups(),
 			},
 		}
-	}()
-	return ch
-}
-
-// runPruneEventTransformer creates a channel for events and
-// starts a goroutine that will read from the channel until it
-// is closed. All events will be republished as Delete events
-// on the provided eventChannel. The function will also return
-// a channel that it will close once the goroutine is shutting
-// down.
-func runPruneEventTransformer(eventChannel chan event.Event) (chan event.Event, <-chan struct{}) {
-	completedChannel := make(chan struct{})
-	tempEventChannel := make(chan event.Event)
-	go func() {
-		defer close(completedChannel)
-		for msg := range tempEventChannel {
-			// If it is not a Prune event, no need to make any transformation.
-			if msg.Type != event.PruneType {
-				eventChannel <- msg
-			} else {
-				eventChannel <- event.Event{
-					Type: event.DeleteType,
-					DeleteEvent: event.DeleteEvent{
-						Operation:  transformPruneOperation(msg.PruneEvent.Operation),
-						Object:     msg.PruneEvent.Object,
-						Identifier: msg.PruneEvent.Identifier,
-						Error:      msg.PruneEvent.Error,
-					},
-				}
-			}
+		// Create a new TaskStatusRunner to execute the taskQueue.
+		klog.V(4).Infoln("destroyer building TaskStatusRunner...")
+		runner := taskrunner.NewTaskStatusRunner(resourceObjs.AllIds(), d.StatusPoller)
+		klog.V(4).Infoln("destroyer running TaskStatusRunner...")
+		err = runner.Run(context.Background(), taskQueue.ToChannel(), eventChannel, taskrunner.Options{
+			UseCache:         true,
+			PollInterval:     defaultPollInterval,
+			EmitStatusEvents: false,
```

Comment:
We should just set this to `true`. Longer-term we should remove this property completely and leave it up to the caller to filter out status events if they are not needed.